ddocs/medic-db/medic-client/views/trainings_by_contact/map.js

```diff
@@ -0,0 +1,5 @@
+function(doc) {
+  if (doc.type === 'data_training' && doc.form && doc.contact) {
+    emit([doc.contact._id], doc.form);
+  }
+}
```

Comment:
1. Do supervisors need to see this data? If not, it'd be better to put it in the user-meta db instead of medic so it doesn't impact other users.
2. For performance reasons it's better to avoid custom views if possible, and instead use CouchDBs built in doc ID view. In this case can we format the ID to make this view unnecessary, for example, if the doc ID was written like "training:<userid>:<formid>" then query using startswith and endswith to get completed trainings for this user?

Reply:
> Do supervisors need to see this data?
Someone from the partner will see it, but I'm not sure the role of this user, I'm going to ask for more details.

> For performance reasons it's better to avoid custom views if possible, and instead use CouchDBs built in doc ID view. In this case can we format the ID to make this view unnecessary, for example, if the doc ID was written like "training::" then query using startswith and endswith to get completed trainings for this user?

1. For my learning. What is a good scenario that justify a custom view?
2. To double check I understand. The ID will be something like this: "training:" + uuid  ---> "training:6b99de2a-dfc4-11ec-9d64-0242ac120002" Is that okay?

Reply:
I remember reading somewhere long ago that using `startswith` and `endswith` to query docs is also expensive. But better than views, right?

Reply:
CouchDB view performance is actually really solid, but it can be slow in creating, or reindexing views. That's why reusing existing views is so powerful - you get all the speed of the read, and none of the cost of the reindexing because it had to do that anyway.

> To double check I understand. The ID will be something like this: "training:" + uuid ---> "training:6b99de2a-dfc4-11ec-9d64-0242ac120002" Is that okay?

OMG! Markdown completely removed part of my comment, sorry! It's supposed to be `training:userid:uuid`. This way you can use startswith and endswith to filter down to just the users docs. Even better if the doc can be stored in the user meta db because then you don't need the `userid` bit.

> I remember reading somewhere long ago that using startswith and endswith to query docs is also expensive. But better than views, right?

startswith and endswith are params to the view query and perform well to get a subset of the view results.

Reply:
Looking across the code I noticed that we use more `userCtx.name`, using it can also improve readability, at the end it turned out: `training:userCtxName:uuid`. I hope this is okay :)

Reply:
Perfect! Good call.